### ISSUE TITLE
Selecting Random Row From Postgres

### ISSUE THREAD

**Issue by anton-mladenov** [typeorm/typeorm#2741]
**Issue type:**

[x ] question
[ ] bug report
[ ] feature request
[ x] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x ] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

I am trying to get a random row from a table in Postgres. 

After reading [this](https://github.com/typeorm/typeorm/issues/813) and [this](https://stackoverflow.com/questions/8674718/best-way-to-select-random-rows-postgresql) I was able to come up with the following piece of code:

```
@Get("/motivations/random")
async getRandomMotivation() {

  return await getConnection()
    .createQueryBuilder()
    .select("id")
    .from(Motivation, "motivation")
    .orderBy("RANDOM()")
    .limit(1)
    .getOne()
}
```

The result is this:

```
query: SELECT id FROM "motivations" "motivation" ORDER BY RANDOM() ASC LIMIT 1
```

However, when sending a GET request with POSTMAN it returns a `Not Found Error` and that's all ...

I was looking in TypeORM's documentation plenty of times, looking for a randomizing solution of any sort but found none.  

Did I miss anything in the query builder? Or did I add something extra? Or just the whole query statement is built in the wrong way? 

Any tip or comment helping solve the problem is welcomed.

**Comment by gintsgints** [comment]
Try check what SQL is produced as result.

**Comment by gintsgints** [comment]
1. Try to execute select right from database tool. Do you get result?
2. Write small test case Without @Get. Just call createQueryBuilder()... and log out result.

**Comment by anton-mladenov** [comment]
@gintsgints here's the latest development ...
The following query inside a `@Get` is logging the random result to the console **but** is not returning it. 
```
@Get("/motivations/random")
async getRandomMotivation() {

const randomFunc = async () => {

    try {
	let entityManager = await Motivation.getRepository()
		.createQueryBuilder()
		.select("motivations.id")
		.from(Motivation, "motivations")
		.orderBy("RANDOM()")
		.limit(1)
		.getOne()

      console.log("__testing: ", entityManager)
      return await entityManager
    }
    catch (error) {
        console.log("___errorrrrr: ", error)
    }

  }

  return await randomFunc()
}
```
When I fire up the server it first returns an `ConnectionNotFoundError: Connection "default" was not found.` error, then it actually loads the server and the database with all tables.
Then, I call the right endpoint, it logs out the random result to the console and when it tries to return the result it returns the following:
```

Connected to Postgres with TypeORM
Listening on port 4000 @ 22:16
query: SELECT "motivations"."id" AS "motivations_id" FROM "motivations" "Motivation", "motivations" "motivations" ORDER BY RANDOM() ASC LIMIT 1
__testing:  Motivation { id: 40 }
query: SELECT "Motivation"."id" AS "Motivation_id", "Motivation"."motivation" AS "Motivation_motivation", "Motivation"."user_id" AS "Motivation_user_id" FROM "motivations" "Motivation" WHERE ("Motivation"."id" = $1) -- PARAMETERS: [null]
query failed: SELECT "Motivation"."id" AS "Motivation_id", "Motivation"."motivation" AS "Motivation_motivation", "Motivation"."user_id" AS "Motivation_user_id" FROM "motivations" "Motivation" WHERE ("Motivation"."id" = $1) -- PARAMETERS: [null]
error: { error: invalid input syntax for integer: "NaN"
.... }
```

As you can see, it runs the random result query and logs it to the console. Then, it runs two queries more with empty parameters ... 

Why TypeORM tries to run two additional queries? How do I overcome this issue?

**Comment by gintsgints** [comment]
@anton-mladenov imho you have issues with asynchronous calls. (it's get called and then some process is done again). As I see from your code you are overusing await.  I would rewrite your code as
```
@Get("/motivations/random")
async getRandomMotivation() {
  try {
	  let record = await Motivation.getRepository()
      .createQueryBuilder()
      .select("motivations.id")
      .from(Motivation, "motivations")
      .orderBy("RANDOM()")
      .limit(1)
      .getOne()

    console.log("__testing: ", record)
    return record
  } catch (error) {
      console.log("___errorrrrr: ", error)
  }
}
```

**Comment by anton-mladenov** [comment]
I found the problem.
I had two paths - `/motivations/:id` and `/motivations/random`.
However, I've forgot to add a check for numbers on `/:id` and because of that whenever I was calling `/random` the server was also calling `/:id` but without any parameter given.
After I changed the path to `/motivations/:id([0-9]+)` the issue is gone. It also turned out that the query to the database was working since the very beginning ... 🤣
